lix/packages/client/src/git-http/optimizeReq.ts

```diff
@@ -0,0 +1,189 @@
+	}: {
+		method: string
+		url: string
+		body?: any
+	}
+) {
+	// Optimize refs requests, GET info/refs?service=git-upload-pack is a request to get the refs but does not support filtering, we rewrite this to a post upload pack v2 request that allows filetering refs how we want
+	// FIXME: document url + method api, how to get tree for commit just filter file blobs
+	// "http://localhost:3001/git-proxy//github.com/inlang/example/info/refs?service=git-upload-pack"
+	if (url.endsWith("info/refs?service=git-upload-pack") && gitConfig.addRefs !== undefined) {
+		const uploadPackUrl = url.replace("info/refs?service=git-upload-pack", "git-upload-pack")
+		const lines = []
+
+		// TODO: #1459 check if we have to ask for the symrefs
+		lines.push(encodePackLine("command=ls-refs"))
+
+		// 0001 - Delimiter Packet (delim-pkt) - separates sections of a message
+		lines.push(encodePackLine("agent=lix") + "0001")
+
+		if (gitConfig.addRefs?.length > 0) {
+			for (let ref of gitConfig.addRefs) {
+				if (!ref.startsWith("refs/")) {
+					ref = "refs/heads/" + ref
+				}
+				lines.push(encodePackLine("ref-prefix " + ref))
+			}
+		}
+
+		lines.push(encodePackLine("ref-prefix HEAD"))
+
+		// TODO: what does this do?
+		lines.push(encodePackLine("symrefs"))
+		// empty line for empty symrefs
+		lines.push(encodePackLine(""))
+
+		return {
+			url: uploadPackUrl,
+			body: lines.join(""),
+			method: "POST",
+			headers: {
+				accept: "application/x-git-upload-pack-result",
+				"content-type": "application/x-git-upload-pack-request",
+				"git-protocol": "version=2",
+			},
+		}
+	}
+
+	// Post requests are always /git-upload-pack requests for the features we currently optimize
+	// && url.endsWith("/git-upload-pack")
+	if (method === "POST") {
+		// decode the lines to be able to change them
+		let rawLines = decodeGitPackLines(body)
+
+		// We modify the raw lines without encoding, we encode them later at once
+		if (gitConfig.noBlobs) {
+			rawLines = addBlobNoneFilter(rawLines)
+		}
+
+		if (gitConfig.overrideHaves) {
+			rawLines = overrideHaves(rawLines, gitConfig.overrideHaves)
+		}
+
+		rawLines = addNoProgress(rawLines)
+
+		// todo add caps helper and always add no progress!
+		if (gitConfig.overrideWants) {
+			rawLines = addWantsCapabilities(rawLines)
+			rawLines = overrideWants(rawLines, gitConfig.overrideWants)
+		}
+
+		// encode lines again to send them in a reques
+		const newBody = rawLines.map((updatedRawLine: any) => encodePackLine(updatedRawLine)).join("")
+
+		return {
+			body: newBody,
+		}
+	}
+
+	return undefined
+}
+
+export async function optimizeRes({
+	origUrl,
+	resBody,
+	statusCode,
+	resHeaders,
+}: {
+	origUrl: string
+	usedUrl: string
+	resBody: Uint8Array
+	statusCode: number
+	resHeaders: Record<string, string>
+}) {
+	// We only optimize ref requests
+	if (!origUrl.endsWith("info/refs?service=git-upload-pack") || statusCode !== 200) {
+		return undefined
+	}
+
+	// TODO: document why we need to override cpabilities
```

Comment:
the capabilities are hardcoded since we don't have them available in the response we get when we replaced the get call with the post to upload pack.